FOUNDATIONS: Add Group Post

ysnarafat opened the pull request:
closes #240

ElbekDeveloper commented on Taarafo.Core.Tests.Unit/Services/Foundations/GroupPosts/GroupPostServiceTests.Exception.Add.cs:
```diff
@@ -0,0 +1,60 @@
+
+namespace Taarafo.Core.Tests.Unit.Services.Foundations.GroupPosts
+{
+    public partial class GroupPostServiceTests
```
Name should be in plural:
Yours: `GroupPostServiceTests.Exception.Add.cs`
Desired: `GroupPostServiceTests.Exceptions.Add.cs`

ElbekDeveloper commented on Taarafo.Core.Tests.Unit/Services/Foundations/GroupPosts/GroupPostServiceTests.Exception.Add.cs:
```diff
@@ -0,0 +1,60 @@
+    public partial class GroupPostServiceTests
+    {
+        [Fact]
+        public async Task ShouldThrowCriticalDependencyExceptionOnAddIfSqlErrorOccursAndLogItAsync()
```
You need to handle a couple more exceptions that EF might throw

ElbekDeveloper commented on Taarafo.Core/Migrations/20220910032910_UpdateEventModel.cs:
```diff
@@ -1,4 +1,9 @@
-﻿using Microsoft.EntityFrameworkCore.Migrations;
+﻿// ---------------------------------------------------------------
```
Good practice would be not to touch any other file that is not in the scope of the issue.

cjdutoit commented on Taarafo.Core/Services/Foundations/GroupPosts/GroupPostService.Validations.cs:
```diff
@@ -0,0 +1,62 @@
+
+            Validate(
+                (Rule: IsInvalid(groupPost.GroupId), Parameter: nameof(GroupPost.GroupId)),
+                (Rule: IsInvalid(groupPost.Group), Parameter: nameof(GroupPost.Group)),
```
Remove this line you only need to validate GroupId and PostId

cjdutoit commented on Taarafo.Core/Services/Foundations/GroupPosts/GroupPostService.Validations.cs:
```diff
@@ -0,0 +1,62 @@
+                (Rule: IsInvalid(groupPost.GroupId), Parameter: nameof(GroupPost.GroupId)),
+                (Rule: IsInvalid(groupPost.Group), Parameter: nameof(GroupPost.Group)),
+                (Rule: IsInvalid(groupPost.PostId), Parameter: nameof(GroupPost.PostId)),
+                (Rule: IsInvalid(groupPost.Post), Parameter: nameof(GroupPost.Post)));
```
Remove this line you only need to validate GroupId and PostId

cjdutoit commented on Taarafo.Core.Tests.Unit/Services/Foundations/GroupPosts/GroupPostServiceTests.Validations.Add.cs:
```diff
@@ -0,0 +1,112 @@
+                key: nameof(GroupPost.GroupId),
+                values: "Id is required");
+
+            invalidGrouPostException.AddData(
```
Remove this, you only need to validate GroupId and PostId

cjdutoit commented on Taarafo.Core.Tests.Unit/Services/Foundations/GroupPosts/GroupPostServiceTests.Validations.Add.cs:
```diff
@@ -0,0 +1,112 @@
+                key: nameof(GroupPost.PostId),
+                values: "Id is required");
+
+            invalidGrouPostException.AddData(
```
Remove this, you only need to validate GroupId and PostId

cjdutoit commented on Taarafo.Core.Tests.Unit/Services/Foundations/GroupPosts/GroupPostServiceTests.cs:
```diff
@@ -0,0 +1,64 @@
+            var filler = new Filler<GroupPost>();
+
+            filler.Setup()
+                .OnType<DateTimeOffset>().IgnoreIt();
```
Add an ignore rule for Group
Add an ignore rule for Post

cjdutoit reviewed (changes requested):
Some small corrections needed.

ElbekDeveloper commented:
This PR is closed due to inactivity

ysnarafat commented:
> This PR is closed due to inactivity

Sorry for the late activity. I have addressed the comments and fixed accordingly. Could you please open the PR again and review the PR?

ElbekDeveloper commented on Taarafo.Core.Tests.Unit/Services/Foundations/GroupPosts/GroupPostServiceTests.Exceptions.Add.cs:
```diff
@@ -0,0 +1,236 @@
+            actualGroupPostDependencyValidationException.Should().BeEquivalentTo(
+                expectedGroupPostDependencyValidationException);
+
+            this.loggingBrokerMock.Verify(broker =>
+                broker.LogError(It.Is(SameExceptionAs(
+                    expectedGroupPostDependencyValidationException))),
+                        Times.Once);
+
+            this.storageBrokerMock.Verify(broker =>
+                broker.InsertGroupPostAsync(someGroupPost),
+                    Times.Once);
+
+            this.loggingBrokerMock.VerifyNoOtherCalls();
```
There is one rule when verifying dependencies. You order these verification statements on the order by which they are called. 
In this case you first call `storageBroker` that throws exception and then `loggingBroker` logs the error. The order of these statements should be the same. This is the correct order: 

``` cs
this.storageBrokerMock.Verify(broker =>
    broker.InsertGroupPostAsync(someGroupPost),
        Times.Once);

this.loggingBrokerMock.Verify(broker =>
    broker.LogError(It.Is(SameExceptionAs(
            expectedGroupPostDependencyValidationException))),
                Times.Once);

this.storageBrokerMock.VerifyNoOtherCalls();
this.loggingBrokerMock.VerifyNoOtherCalls();

```

ysnarafat replied:
Okay, correcting according to this way.
